### PR TITLE
Migrating sources from dotnet/extensions

### PR DESCRIPTION
**Pull request by JunTaoLuo**
This PR is made on top of https://github.com/dotnet/aspnetcore/pull/19032 to simplify reviews. When approved, the changes in both PRs will be merged into master.

Update: all build/test issues have been resolved and I have re-targeted the base branch to master, I'll add more details in the PR description but this is ready to be merged.

Components being migrated:
- M.E.FileProviders.Embedded
- M.E.FileProviders.Embedded.Manifest.Task
- M.E.Configuration.KeyPerFile
- M.E.Localization
- M.E.Localization.Abstraction
- M.E.ObjectPool
- M.JSInterop
- M.JSInterop.JS
- M.E.WebEncoders
- M.E.Diagnostics.HealthChecks
- M.E.Diagnostics.HealthChecks.Abstractions

Note that I anticipate that we will need to address a few more issues in collaboration with the dotnet/runtime team. Specifically:
- How to consolidate our usages of shared sources
- How to split up Logging.Testing
- How to transport reference assemblies from dotnet/runtime

I'll work with @maryamariyan on the dotnet side to address the remaining concerns once there is progress on that side.

### CODE REVIEW

File: eng/SharedFramework.External.props [hunk 1]
```diff
@@ -7,7 +7,7 @@
 <Project>
 
   <ItemGroup>
-    <!-- Dependencies from dotnet/extensions -->
+    <!-- Dependencies from dotnet/runtime -->
```
Comment (JunTaoLuo) [inline eng/SharedFramework.External.props]: may be a little premature.

File: eng/targets/ResolveReferences.targets [hunk 2]
```diff
@@ -119,7 +119,7 @@
     <Error Condition="@(_InvalidReferenceToSharedFxOnlyAssembly->Count()) != 0"
            Text="Cannot reference &quot;%(_InvalidReferenceToSharedFxOnlyAssembly.Identity)&quot; directly because it is part of the shared framework and this project is not. Use &lt;FrameworkReference Include=&quot;Microsoft.AspNetCore.App&quot; /&gt; instead." />
 
-    <Error Condition="@(_InvalidReferenceToNonSharedFxAssembly->Count()) != 0"
+    <Error Condition="@(_InvalidReferenceToNonSharedFxAssembly->Count()) != 0 AND '$(TargetFramework)' == '$(DefaultNetCoreTargetFramework)'"
```
Comment (JunTaoLuo) [inline eng/targets/ResolveReferences.targets]: The issue here is that there are certain netstandard2.0 dependencies in our extensions packages that were flagged by this logic. We should only be concerned with netcoreapp5.0 dependencies.

File: src/Components/Web.JS/dist/Release/blazor.server.js [hunk 3]
```diff
@@ -12,4 +12,4 @@ var r=n(50),o=n(51),i=n(52);function a(){return c.TYPED_ARRAY_SUPPORT?2147483647
...
...lice())},e.prototype.writeStreamInvocation=function(e){var t=o().encode([i.MessageType.StreamInvocation,e.headers||{},e.invocationId,e.target,e.arguments,e.streamIds]);return a.write(t.slice())},e.prototype.writeStreamItem=function(e){var t=o().encode([i.MessageType.StreamItem,e.headers||{},e.invocationId,e.item]);return a.write(t.slice())},e.prototype.writeCompletion=function(e){var t,n=o(),r=e.error?this.errorResult:e.result?this.nonVoidResult:this.voidResult;switch(r){case this.errorResult:t=n.encode([i.MessageType.Completion,e.headers||{},e.invocationId,r,e.error]);break;case this.voidResult:t=n.encode([i.MessageType.Completion,e.headers||{},e.invocationId,r]);break;case this.nonVoidResult:t=n.encode([i.MessageType.Completion,e.headers||{},e.invocationId,r,e.result])}return a.write(t.slice())},e.prototype.writeCancelInvocation=function(e){var t=o().encode([i.MessageType.CancelInvocation,e.headers||{},e.invocationId]);return a.write(t.slice())},e.prototype.readHeaders=function(e){var t=e[1];if("object"!=typeof t)throw new Error("Invalid headers.");return t},e}();n.d(t,"VERSION",function(){return u}),n.d(t,"MessagePackHubProtocol",function(){return c});var u="0.0.0-DEV_BUILD"}]);
```
Comment (JunTaoLuo) [inline src/Components/Web.JS/dist/Release/blazor.server.js]: Need to revert

File: src/Configuration/Config.KeyPerFile/test/ConfigurationProviderTestBase.cs [hunk 4]
```diff
@@ -0,0 +1,761 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (JunTaoLuo) [inline src/Configuration/Config.KeyPerFile/test/ConfigurationProviderTestBase.cs]: Considering doing this change in the migrations PR instead.

File: src/FileProviders/Manifest.MSBuildTask/src/Manifest.cs [hunk 5]
```diff
@@ -11,7 +11,9 @@ namespace Microsoft.Extensions.FileProviders.Embedded.Manifest.Task
 {
     public class Manifest
     {
+#pragma warning disable PUB0001 // Pubternal type IActivator in public API
```
Comment (JunTaoLuo) [inline src/FileProviders/Manifest.MSBuildTask/src/Manifest.cs]: This pubternal doesn't matter as much given that it will be compiled into a msbuild task dll. I think keeping the warning disable is fine.

File: src/Localization/Localization/src/ResourceManagerStringLocalizer.cs [hunk 6]
```diff
@@ -54,7 +54,9 @@ public class ResourceManagerStringLocalizer : IStringLocalizer
         /// </summary>
         public ResourceManagerStringLocalizer(
             ResourceManager resourceManager,
+#pragma warning disable PUB0001 // Pubternal type IActivator in public API
```
Comment (JunTaoLuo) [inline src/Localization/Localization/src/ResourceManagerStringLocalizer.cs]: @mkArtakMSFT who should I talk to regarding these pubternal APIs in Localization?

File: src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.csproj [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <!-- This file is automatically generated. -->
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net46</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
```
Comment (dougbu) [inline src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.csproj]: Why is there a ref/ project for Microsoft.AspNetCore.Testing at all? Should see warnings about non-`$(IsAspNetCoreApp)` projects having ref/ projects (though those are likely buried in other warnings at the moment)

File: src/Components/Components/ref/Microsoft.AspNetCore.Components.csproj [hunk 8]
```diff
@@ -7,13 +7,11 @@
     <Compile Include="Microsoft.AspNetCore.Components.netstandard2.0.cs" />
     <Reference Include="Microsoft.Extensions.Logging.Abstractions"  />
     <Reference Include="Microsoft.Extensions.DependencyInjection.Abstractions"  />
-    <Reference Include="Microsoft.Extensions.HashCodeCombiner.Sources"  />
```
Comment (dougbu) [inline src/Components/Components/ref/Microsoft.AspNetCore.Components.csproj]: Need to update the *.Manual.cs files. As-is I expect compilation errors due to missing `internal` classes for some (if not many) uses of the ref/ assemblies

File: eng/Version.Details.xml [hunk 9]
```diff
@@ -420,11 +352,7 @@
       <Uri>https://github.com/dotnet/arcade</Uri>
       <Sha>b0e8d3944155f94f83deea8afe025debe369e69f</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.AspNetCore.Testing" Version="5.0.0-preview.1.20112.1" CoherentParentDependency="Microsoft.AspNetCore.Razor.Language">
-      <Uri>https://github.com/dotnet/extensions</Uri>
-      <Sha>5fe6ff74c3f95683d21e33cd6bf98e8e1477d6ed</Sha>
-    </Dependency>
-    <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="3.5.0-beta3-20111-03" CoherentParentDependency="Microsoft.Extensions.Logging">
+    <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="3.5.0-beta3-20112-03" CoherentParentDependency="Microsoft.Extensions.Logging">
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply eng/Version.Details.xml]: Bad rebase, will fix after I change a bit of the layout in the base branch

File: src/Localization/Localization/src/ResourceManagerStringLocalizer.cs (hunk 6 above)
Comment (ryanbrandenburg) [reply src/Localization/Localization/src/ResourceManagerStringLocalizer.cs]: Is the idea here that these constructors and their PubTernal types should become internal and the tests should use IVT? That seems reasonable, file an issue and we can go from there.

The pragma comments seem to be referring to an `IActivator` type which is not actually used here though.

Comment (JunTaoLuo) [reply src/Localization/Localization/src/ResourceManagerStringLocalizer.cs]: The comments are my bad, copy-pasta. The question is how should we deal with the pubternal types. Should we make the types/methods public? Or should we make them internal? We did a pass in aspnetcore to remove pubternals but it seems like we didn't do so for extensions. We just need to go through the same process again for Localization.

Comment (dougbu) [reply src/Localization/Localization/src/ResourceManagerStringLocalizer.cs]: Suggest filing an issue describing the proposed changes in detail e.g. "burn pub-ternal with fire and make them all `internal`" or whatever makes sense. Then, mark for API review.

File: eng/targets/ResolveReferences.targets (hunk 2 above)
Comment (JunTaoLuo) [reply eng/targets/ResolveReferences.targets]: There are two projects not multitargeted:
- Microsoft.Extensions.FileProviders.Embedded.Manifest.Task, but it builds a msbuild task assembly so it's fine
- Microsoft.Extensions.ObjectPool, it doesn't have any dependencies so it won't trigger this check. However, I still think we probably want to multitarget it and it was an oversight that it wasn't.

File: eng/Build.props [hunk 10]
```diff
@@ -151,6 +151,14 @@
...
                           $(RepoRoot)src\Components\**\*.csproj;
                           $(RepoRoot)src\Analyzers\**\*.csproj;
+                          $(RepoRoot)src\FileProviders\**\*.csproj;
+                          $(RepoRoot)src\Configuration\**\*.csproj;
+                          $(RepoRoot)src\Localization\**\*.csproj;
+                          $(RepoRoot)src\ObjectPool\**\*.csproj;
+                          $(RepoRoot)src\JSInterop\**\*.csproj;
+                          $(RepoRoot)src\WebEncoders\**\*.csproj;
+                          $(RepoRoot)src\HealthChecks\**\*.csproj;
+                          $(RepoRoot)src\TestingUtils\**\*.csproj;
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply eng/Build.props]: I don't see much benefit.

File: eng/Build.props [hunk 11]
```diff
@@ -151,6 +151,14 @@
                           $(RepoRoot)src\SignalR\**\*.csproj;
                           $(RepoRoot)src\Components\**\*.csproj;
                           $(RepoRoot)src\Analyzers\**\*.csproj;
+                          $(RepoRoot)src\FileProviders\**\*.csproj;
+                          $(RepoRoot)src\Configuration\**\*.csproj;
+                          $(RepoRoot)src\Localization\**\*.csproj;
+                          $(RepoRoot)src\ObjectPool\**\*.csproj;
+                          $(RepoRoot)src\JSInterop\**\*.csproj;
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply eng/Build.props]: Added to npm projects

File: src/FileProviders/Manifest.MSBuildTask/src/Manifest.cs (hunk 5 above)
Comment (JunTaoLuo) [reply src/FileProviders/Manifest.MSBuildTask/src/Manifest.cs]: Actually on second thought, there's no reason why this needs to be pubternal. I'm going to make it public since this package isn't shipped so there's no harm in exposing the API.

File: src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.csproj (hunk 7 above)
Comment (JunTaoLuo) [reply src/TestingUtils/Microsoft.AspNetCore.Testing/ref/Microsoft.AspNetCore.Testing.csproj]: Removed

File: src/Localization/Localization/src/ResourceManagerStringLocalizer.cs (hunk 6 above)
Comment (JunTaoLuo) [reply src/Localization/Localization/src/ResourceManagerStringLocalizer.cs]: Filed https://github.com/dotnet/aspnetcore/issues/19081

File: src/FileProviders/Embedded/src/Microsoft.Extensions.FileProviders.Embedded.csproj [hunk 12]
```diff
@@ -9,6 +9,10 @@
     <NuspecFile Condition="'$(DotNetBuildFromSource)' == 'true'">$(MSBuildProjectName).netcoreapp.nuspec</NuspecFile>
     <IsPackable>true</IsPackable>
     <IsShipping>true</IsShipping>
+    <IsAspNetCoreApp>true</IsAspNetCoreApp>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>files;filesystem</PackageTags>
+    <NoWarn>$(NoWarn);NU5100</NoWarn>
```
Comment (JunTaoLuo) [inline src/FileProviders/Embedded/src/Microsoft.Extensions.FileProviders.Embedded.csproj]: I see a warning
```
error NU5100: The assembly 'tasks\netstandard2.0\Microsoft.Extensions.FileProviders.Embedded.Manifest.Task.dll' is not inside the 'lib' folder and hence it won't be added as a reference when the package is installed into a project. Move it into the 'lib' folder if it needs to be referenced. [C:\gh\aspnetcore2\src\FileProviders\Embedded\src\Microsoft.Extensions.FileProviders.Embedded.csproj]
```
This seems to only occur in aspnetcore builds and not in extensions. Perhaps we have a different sdk?

**Comment by JunTaoLuo** [comment]
@dotnet/aspnet-build This is ready for review now.

File: src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj [hunk 13]
```diff
@@ -18,13 +18,15 @@
     <Compile Include="$(SharedSourceRoot)HttpSys\**\*.cs" Link="Shared\HttpSys\%(Filename)%(Extension)"/>
     <Compile Include="$(SharedSourceRoot)NonCapturingTimer\*.cs" Link="Shared\NonCapturingTimer\%(Filename)%(Extension)"/>
     <Compile Include="$(SharedSourceRoot)ObjectMethodExecutor\*.cs" Link="Shared\ObjectMethodExecutor\%(Filename)%(Extension)"/>
+    <Compile Include="$(SharedSourceRoot)ParameterDefaultValue\*.cs" Link="Shared\ParameterDefaultValue\%(Filename)%(Extension)"/>
```
Comment (dougbu) [inline src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj]: Why is this needed?

File: eng/targets/ResolveReferences.targets (hunk 2 above)
Comment (dougbu) [reply eng/targets/ResolveReferences.targets]: Change Microsoft.Extensions.ObjectPool so that it is multi-targeted.

**Review by dougbu — CHANGES_REQUESTED** [review]

File: eng/Workarounds.FileProviders.Embedded.props [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<!-- Use this file to workaround issues with referencing M.E.FileProviders.Embedded. List the issue tracking the item to fix so we can remove the workaround when the issue is resolved. -->
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply eng/Workarounds.FileProviders.Embedded.props]: I agree. Originally, I expected the workaround to be applicable to more than one project. But I think it's sufficient to only apply it to Components.Server.

File: src/Components/Components/ref/Microsoft.AspNetCore.Components.csproj (hunk 8 above)
Comment (JunTaoLuo) [reply src/Components/Components/ref/Microsoft.AspNetCore.Components.csproj]: The compilations errors have been resolved.

File: src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj (hunk 13 above)
Comment (JunTaoLuo) [reply src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj]: Not needed, it was mistakenly included in the merge between extensions and aspnetcore during migration.

File: eng/targets/ResolveReferences.targets (hunk 2 above)
Comment (JunTaoLuo) [reply eng/targets/ResolveReferences.targets]: Done

File: src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj (hunk 13 above)
Comment (JunTaoLuo) [reply src/Shared/test/Shared.Tests/Microsoft.AspNetCore.Shared.Tests.csproj]: Correction, it is needed to compile ActivatorUtilities tests.

File: src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj [hunk 15]
```diff
@@ -39,6 +40,15 @@
...
   </PropertyGroup>
 
+  <!-- This workaround is required when referencing FileProviders.Embedded as a project -->
+  <PropertyGroup>
+    <_FileProviderTaskAssembly>$(ArtifactsDir)bin\Microsoft.Extensions.FileProviders.Embedded.Manifest.Task\$(Configuration)\netstandard2.0\Microsoft.Extensions.FileProviders.Embedded.Manifest.Task.dll</_FileProviderTaskAssembly>
+  </PropertyGroup>
+
+  <UsingTask
+    AssemblyFile="$(_FileProviderTaskAssembly)"
+    TaskName="Microsoft.Extensions.FileProviders.Embedded.Manifest.Task.GenerateEmbeddedResourcesManifest" />
```
Comment (dougbu) [inline src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj]: nit: Doesn't really matter but the alternative would be to `<import />` the .props file here (after setting `$(_FileProviderTaskAssembly)`)

File: src/Components/Server/src/Directory.Build.targets [hunk 16]
```diff
@@ -0,0 +1,6 @@
+﻿<Project>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\, Directory.Build.targets))\Directory.Build.targets" />
+  <!-- Workaround target import for project references to Microsoft.Extensions.FileProviders.Embedded -->
+  <Import
+    Project="$(RepoRoot)src\FileProviders\Embedded\src\build\netstandard2.0\Microsoft.Extensions.FileProviders.Embedded.targets" />
```
Comment (dougbu) [inline src/Components/Server/src/Directory.Build.targets]: If this works, :ok: I guess. But my preference would be to do this `<Import />` in the project file rather than adding this additional file

File: src/ObjectPool/src/Microsoft.Extensions.ObjectPool.csproj [hunk 17]
```diff
@@ -2,12 +2,13 @@
 
   <PropertyGroup>
     <Description>A simple object pool implementation.</Description>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>netstandard2.0;$(DefaultNetCoreTargetFramework)</TargetFrameworks>
```
Comment (dougbu) [inline src/ObjectPool/src/Microsoft.Extensions.ObjectPool.csproj]: 👍

**Review by dougbu — APPROVED** [review]

File: src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj (hunk 15 above)
Comment (JunTaoLuo) [reply src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj]: If we did that then the `_FileProviderTaskAssembly` property would be overridden by the imported props.

File: src/Components/Server/src/Directory.Build.targets (hunk 16 above)
Comment (JunTaoLuo) [reply src/Components/Server/src/Directory.Build.targets]: I tried that but it doesn't work in the project file since the `PrepareResourceNamesDependsOn` property will be overridden by the SDK.

**Comment by JunTaoLuo** [comment]
Waiting for https://github.com/dotnet/aspnetcore/pull/19122

**Comment by JunTaoLuo** [comment]
@javiercn @SteveSandersonMS who can help me take a look at the Blazor Template tests failures? I can't seem to run the tests locally so I can't debug the failure described in the CI test logs.

**Comment by JunTaoLuo** [comment]
It turns out there are binlogs captured on the CI, so I might be unblocked.

**Comment by SteveSandersonMS** [comment]
@JunTaoLuo Please let us know if you still need assistance after checking those logs.

Today was the final day before code-complete for the current Blazor milestone, so Javier and I were working pretty furiously on that, hence not responding sooner. Hope that's OK!

**Comment by JunTaoLuo** [comment]
Turns out it's an issue with ref packs again. I'll look into a fix. This PR might need some additional verification.

**Comment by JunTaoLuo** [comment]
@dotnet/aspnet-build can y'all look at this again before I merge? I'll also try to compare the ref pack outputs to ensure we didn't miss anything.

**Comment by JunTaoLuo** [comment]
I verified the ref packs and everything looks good

File: eng/Dependencies.props [hunk 18]
```diff
@@ -94,14 +78,17 @@ and are generated based on the last package release.
     <LatestPackageReference Include="System.Net.Http" Version="$(SystemNetHttpPackageVersion)" />
     <LatestPackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataPackageVersion)" />
     <LatestPackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="$(SystemRuntimeCompilerServicesUnsafePackageVersion)" />
+    <LatestPackageReference Include="System.Runtime.InteropServices.RuntimeInformation" Version="$(SystemRuntimeInteropServicesRuntimeInformationPackageVersion)" />
```
Comment (wtgodbe) [inline eng/Dependencies.props]: For my understanding - these were packages that (previously) Extensions depended on, but not this repo? And now we're adding a dependency on them since we're bringing in the Extensions stuff that uses them?

Comment (JunTaoLuo) [reply eng/Dependencies.props]: Yea

**Review by wtgodbe — APPROVED** [review]
Infra changes look reasonable - :shipit:
